### PR TITLE
[Jetpack Focus] Revamp Landing Screen - Sensor animation

### LINKED ISSUE #17075: [Revamp Landing Screen] JPAndroid - Fine-tune the animation to respond to sensors data when available
Subtly influence the [large text animation](https://github.com/wordpress-mobile/WordPress-Android/issues/17074) on the landing screen of the Jetpack Android app based on sensor data when available.

📜 We're still deciding how the sensor should affect the animation.

### PR DESCRIPTION
**Pull request by ovitrif**
Fixes #17075

With this PR we're using accelerometer sensor data (y-axis acceleration) to influence the speed of the vertical scrolling text from the background of the Jetpack App Landing Screen.

~~⚠️ **Important Notice** This PR should target `trunk` after `feature/revamp-landing-screen` is merged.
PR #17203 should be merged first, then we can target `trunk` with this PR and set it `Ready for review`.~~

To test:

<details>
<summary><b>Prerequisite</b> - Enable the <code>LandingScreenRevamp</code> feature flag</summary>

- Either `login` → `Me` → `App Settings` → `Debug Settings` → enable `LandingScreenRevamp` → `RESTART THE APP` → `logout` or
- Apply this hard-coded solution:

<details>
<summary>Patch to enable the <code>LANDING_SCREEN_REVAMP</code> build flag</summary>

```diff
diff --git a/WordPress/build.gradle b/WordPress/build.gradle
index 63260a4ca8..fe999fa92f 100644
--- a/WordPress/build.gradle
+++ b/WordPress/build.gradle
@@ -112,7 +112,7 @@ android {
         buildConfigField "boolean", "READER_COMMENTS_MODERATION", "false"
         buildConfigField "boolean", "SITE_INTENT_QUESTION", "true"
         buildConfigField "boolean", "SITE_NAME", "false"
-        buildConfigField "boolean", "LANDING_SCREEN_REVAMP", "false"
+        buildConfigField "boolean", "LANDING_SCREEN_REVAMP", "true"
         buildConfigField "boolean", "LAND_ON_THE_EDITOR", "false"
         buildConfigField "boolean", "BLOGGING_PROMPTS", "false"
         buildConfigField "boolean", "QUICK_START_EXISTING_USERS_V2", "false"
```  
</details>

---

</details>

##### 1️⃣ Test Case 1: Happy Flow

0. Open the JP app and make sure you're logged out
1. Make sure you're holding the device in your hand slightly rotated vertically, as illustrated:
   
3. **Expect** the large text on the new login screen to scroll down by default at an accelerated speed
4. Lay the device flat on your office table
5. **Expect** The text to scroll **down** at a slower speed
6. Pick up the device and rotate it slightly inverse than in the illustration (top of the phone should be below its bottom)
7. **Expect** The text to scroll **up**

##### 2️⃣ Test Case 2: Feature flag off

0. Disable the <code>LandingScreenRevamp</code> feature flag (revert change from **Prerequisite**)
1. Open the JP app and make sure you're logged out
2. **Expect** To be presented with the previous design
3. Tap any of the 2 buttons
4. **Expect** the Login flows to still work as usual.

##### 3️⃣ Test Case 3: Sensors off

1. Enable developer mode:
   `Settings` → `About phone / emulated device` → tap `Build number` 7 times
2. Enable `Sensors Off quick setting:
   `Go to `Settings` → `System` → `Developer options` → `Quick settings developer tiles` tap `Sensors Off`
3. Toggle Sensors off:
   Swipe down from the top of the screen and tap `Sensors Off`
4. Open the JP app and make sure you're logged out
5. **Expect** the large text to scroll down at a speed similar to the normal position speed when sensors are enabled
   (step 3 of Test Case 1)

##### 4️⃣ Test Case 4: No accelerometer sensor

1. Apply this patch to simulate missing accelerometer sensor on the device:

```diff
diff --git a/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt b/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
index 99e8570448..c57f1aaf1e 100644
--- a/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
+++ b/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
@@ -62,7 +62,7 @@ class LoginPrologueRevampedViewModel @Inject constructor(
 
         override fun onActive() {
             super.onActive()
-            val sensor = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
+            val sensor: Sensor? = null
             sensorManager.registerListener(this, sensor, SensorManager.SENSOR_DELAY_GAME)
         }
```

2. Open the JP app and make sure you're logged out

3. **Expect** the large text to scroll down at a speed similar to the normal position speed when sensors are enabled

<details>
<summary><h4>🎬 Video Preview</h4></summary>

https://user-images.githubusercontent.com/4588074/192004342-b2955e14-95b0-45d3-87ff-9db4f5ccce28.mp4

- Initially the device is rotated to `-45°` (reflecting a normal use position)
- At second `5` the device is rotated to `-90°` (reflecting flat on table position)
- At second `13` the device is rotated to `-105°` (inverted rotation that triggers reverse scrolling)

---

</details>

## Regression Notes
4. Potential unintended areas of impact
   New Jetpack Landing Screen when sensor is not available.

8. What I did to test those areas of impact (or what existing automated tests I relied on)
   Tested manually different configs: sensors disabled / missing and added test cases for these.

9. What automated tests I added (or what prevented me from doing so)
   None, we're interacting with platform code and UI animation code.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by antonis — COMMENTED** [review]
Great work @ovitrif 👍 

I started reviewing the changes and testing the app. I noticed that the text moves too fast on the initial position. This is more of a personal opinion since I may have not followed all the conversations on this but wanted to confirm that what I'm reproducing on Pixel 5 (Android 13) below is the expected behaviour or the speed may be different per device. In the video I started with my normal holding position (~ 60 degrees) and ended with the device flat on my desk. 

https://user-images.githubusercontent.com/304044/192317085-525f8f05-e74b-4a97-b199-e28fd2ce3662.mp4

I'll continue the review and get back to you.

**Comment by ovitrif** [comment]
> I started reviewing the changes and testing the app. I noticed that the text moves too fast on the initial position. This is more of a personal opinion since I may have not followed all the conversations on this but wanted to confirm that what I'm reproducing on Pixel 5 (Android 13) below is the expected behaviour or the speed may be different per device. In the video I started with my normal holding position (~ 60 degrees) and ended with the device flat on my desk.

Your point is perfectly valid to me @antonis, I haven't tweaked the speed on top of @mkevins implementation with Accelerometer sensor which is the source of this PR — therefore the credit should go to Matthew for this great work 🚀 — but we were discussing this last week.

I wanted to get early feedback and proceed to make adjustments with a more informed stance, so this worked quite great, thank you for sharing your view 👍

I'll look into options to tweak the speed for the most expected / normal phone vertical orientation.

**Comment by ovitrif** [comment]
> I'll look into options to tweak the speed for the most expected / normal phone vertical orientation.

I [decreased the maximum velocity](https://github.com/wordpress-mobile/WordPress-Android/pull/17207/commits/4e5059ff3cd8ed6d81505395c2d3cd189870b67b) — thus obtaining a slower speed @ normal vertical orientation
I also tested the scenario with sensors off and `null` (it required a small fix — a default value for the `acceleration`).

@mkevins Let me know if [this change](https://github.com/wordpress-mobile/WordPress-Android/pull/17207/commits/2817914dbf06b0a2a2cc00dac84050c1ecbd52b9) is aligned with the model 🙇 

P.S. From my side I think this PR is ready for review, the only reason it's a draft is because https://github.com/wordpress-mobile/WordPress-Android/pull/17203 should be merged first into `trunk`, and then this one should target `trunk` 🎉 .

P.P.S. Added Test Case 3️⃣ for testing the animation when sensors are disabled.

**Comment by mkevins** [comment]
> @mkevins Let me know if [this change](https://github.com/wordpress-mobile/WordPress-Android/pull/17207/commits/2817914dbf06b0a2a2cc00dac84050c1ecbd52b9) is aligned with the model 🙇

Indeed this is looking good. Thanks for the fix for the non-sensor version. The reason it wasn't working well, after testing on an emulator without sensors is that the `DRIFT` was being applied in the `onSensorChanged` function instead of inside the `updateForFrame` function. This meant that the `DRIFT` was essentially absent from devices without sensor data on.

The following patch demonstrates a more appropriate place for the `DRIFT` to be applied:

<details>
<summary>Patch to apply DRIFT correctly</summary>

```patch
diff --git a/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt b/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
index 99e8570448..3e69a75d12 100644
--- a/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
+++ b/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
@@ -31,7 +31,7 @@ private const val DRIFT = -0.05f
 class LoginPrologueRevampedViewModel @Inject constructor(
     @ApplicationContext appContext: Context,
 ) : ViewModel() {
-    private var acceleration = -0.3f // Default value when sensor is off
+    private var acceleration = 0f
     private var velocity = 0f
     private var position = 0f
 
@@ -47,7 +47,7 @@ class LoginPrologueRevampedViewModel @Inject constructor(
      */
     fun updateForFrame(elapsed: Float) {
         // Update the velocity, (decayed and clamped to the maximum)
-        velocity = (velocity * exp(elapsed * VELOCITY_DECAY) + elapsed * acceleration)
+        velocity = (velocity * exp(elapsed * VELOCITY_DECAY) + elapsed * (acceleration + DRIFT))
                 .coerceIn(-MAXIMUM_VELOCITY, MAXIMUM_VELOCITY)
         // Update the position, modulo 1 (ensuring a value greater or equal to 0, and less than 1)
         position = ((position + elapsed * velocity) % 1 + 1) % 1
@@ -73,7 +73,7 @@ class LoginPrologueRevampedViewModel @Inject constructor(
 
         override fun onSensorChanged(event: SensorEvent?) {
             event?.values?.let { (_, yAxisAcceleration, _) ->
-                acceleration = yAxisAcceleration * ACCELERATION_FACTOR + DRIFT
+                acceleration = yAxisAcceleration * ACCELERATION_FACTOR
                 postValue(position)
             }
         }
```
</details>

However, the `DRIFT` may have a slightly different purpose than the default / initial acceleration value. The `DRIFT` variable biases the acceleration such that a device  _with_ sensor data on will still result in scrolling text when the device is flat on a table. The value of this bias influences the behavior of the interactive animation in an ongoing way.

So, the way that you have addressed the "sensor-less" case has the advantage that we can configure different scrolling speeds for devices without sensors and for devices with sensors that happen to be flat on a table. For example, we may want that "flat on a table" case to be _slower_ than the default "sensor-less" case. I hope this makes sense. tl;dr, I think your change to the default acceleration is a great solution 😄

File: WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedFragment.kt [hunk 1]
```diff
@@ -64,6 +77,35 @@ class LoginPrologueRevampedFragment : Fragment() {
...
+private fun PositionProvider(
+    viewModel: LoginPrologueRevampedViewModel = viewModel(),
+    content: @Composable () -> Unit
+) {
+    val position = viewModel.positionData.observeAsState(0f)
+    CompositionLocalProvider(LocalPosition provides position.value) {
+        LaunchedEffect(Unit) {
+            var lastFrameNanos: Long? = null
+            while (isActive) {
+                val currentFrameNanos = awaitFrame()
```
Comment (ovitrif) [inline WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedFragment.kt]: @mkevins I've found this issue in the Android Public Tracker:
[Add a lint check for using awaitFrame inside Compose code 
](https://issuetracker.google.com/issues/229748038)

It makes an argument on why we shouldn't use `awaitFrame` (hint: I was researching `awaitFrame`) in Compose code, but rather the `withFrameNanos` from Compose's own time source: [MonotonicFrameClock](https://developer.android.com/reference/kotlin/androidx/compose/runtime/MonotonicFrameClock#withFrameNanos(kotlin.Function1)).

I will adapt the code to use that instead and test if it still works as expected

Comment (ovitrif) [reply WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedFragment.kt]: Update: Still works nicely, I applied the change in https://github.com/wordpress-mobile/WordPress-Android/pull/17207/commits/20e21a4b3964054f2898e56e0833cdf53249a6b1

Comment (mkevins) [reply WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedFragment.kt]: Oh nice find Ovi! This is a good improvement that will helpful for supporting possible future UI testing 👍

**Review by antonis — APPROVED** [review]
Great work @ovitrif 👍 
I've retested with the latest changes and everything worked great for me. The code also LGTM 🎉 

Note: I'm still able to reproduce the bars issue [but this seems fixed in the next design tweaks PR.](https://github.com/wordpress-mobile/WordPress-Android/pull/17231#issuecomment-1262207344)

**Comment by ovitrif** [comment]
> Great work @ovitrif 👍 I've retested with the latest changes and everything worked great for me. The code also LGTM 🎉
> 
> Note: I'm still able to reproduce the bars issue [but this seems fixed in the next design tweaks PR.](https://github.com/wordpress-mobile/WordPress-Android/pull/17231#issuecomment-1262207344)

Thank you for the review @antonis 🙇 

I've created a separate ticket on the project board to test more intensively the issue with the system bars.

Let's give it another testing round after merging the remaining implementation work, when working on that task 👍
